src/round.rs

```diff
@@ -85,14 +92,155 @@ fn span_for_digits(digits: u16) -> u32 {
+    /// let dt = Utc.ymd(2018, 1, 11).and_hms_milli(12, 0, 0, 154);
+    /// assert_eq!(
+    ///     dt.duration_trunc(Duration::milliseconds(10)).unwrap().to_string(),
+    ///     "2018-01-11 12:00:00.150 UTC"
+    /// );
+    /// assert_eq!(
+    ///     dt.duration_trunc(Duration::days(1)).unwrap().to_string(),
+    ///     "2018-01-11 00:00:00 UTC"
+    /// );
+    /// ```
+    fn duration_trunc(self, duration: Duration) -> Result<Self, Self::Err>;
+}
+
+/// The maximum number of seconds a DateTime can be to be represented as nanoseconds
+const MAX_SECONDS_TIMESTAMP_FOR_NANOS: i64 = 9_223_372_036;
+
+impl<Tz: TimeZone> DurationRound for DateTime<Tz> {
+    type Err = RoundingError;
+
+    fn duration_round(self, duration: Duration) -> Result<Self, Self::Err> {
+        if let Some(span) = duration.num_nanoseconds() {
+            if self.timestamp().abs() > MAX_SECONDS_TIMESTAMP_FOR_NANOS {
+                return Err(RoundingError::TimestampExceedsLimit);
+            }
+            let stamp = self.timestamp_nanos();
+            if span > stamp.abs() {
+                return Err(RoundingError::DurationExceedsTimestamp);
+            }
+            let delta_down = stamp % span;
+            if delta_down == 0 {
+                Ok(self)
+            } else {
+                let (delta_up, delta_down) = if delta_down < 0 {
+                    (delta_down.abs(), span - delta_down.abs())
+                } else {
+                    (span - delta_down, delta_down)
+                };
+                if delta_up <= delta_down {
+                    Ok(self + Duration::nanoseconds(delta_up))
+                } else {
+                    Ok(self - Duration::nanoseconds(delta_down))
+                }
+            }
+        } else {
+            Err(RoundingError::DurationExceedsLimit)
+        }
+    }
+
+    fn duration_trunc(self, duration: Duration) -> Result<Self, Self::Err> {
+        if let Some(span) = duration.num_nanoseconds() {
+            if self.timestamp().abs() > 9_223_372_036 {
+                return Err(RoundingError::TimestampExceedsLimit);
+            }
+            let stamp = self.timestamp_nanos();
+            if span > stamp.abs() {
+                return Err(RoundingError::DurationExceedsTimestamp);
+            }
+            let delta_down = stamp % span;
+            match delta_down.cmp(&0) {
+                Ordering::Equal => Ok(self),
+                Ordering::Greater => Ok(self - Duration::nanoseconds(delta_down)),
+                Ordering::Less => Ok(self - Duration::nanoseconds(span - delta_down.abs())),
+            }
+        } else {
+            Err(RoundingError::DurationExceedsLimit)
+        }
+    }
+}
+
+/// An error from rounding by `Duration`
+///
+/// See: [`DurationRound`]
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum RoundingError {
+    DurationExceedsTimestamp,
+    DurationExceedsLimit,
+    TimestampExceedsLimit,
```

Comment:
Could you add docs/examples for these? You can feel free to move the tests for them up into the docstrings.